textarea.go

```diff
@@ -0,0 +1,465 @@
+	isOutsideBuffer := false
+	for i := len(indexes) - 1; i >= 0; i-- {
+		if indexes[i].Line != bufferLine {
+			continue
+		}
+		pos := len([]rune(buffers[bufferLine][:indexes[i].Pos]))
+		if pos <= bufferPosition {
+			indexLine = i
+			if size := len([]rune(buffers[bufferLine])); size < bufferPosition {
+				bufferPosition = size
+				isOutsideBuffer = true
+			}
+			indexPos = bufferPosition - pos
+			break
+		}
+	}
+	if indexLine < 0 {
+		// TODO: find that situation
+		indexLine = 0
+	}
+	if indexPos < 0 {
+		// TODO: find that situation
+		indexPos = 0
+	}
+
+	// convert position from indexes to grapheme for cursor
+	var posInGrapheme int
+	{
+		buf := buffers[indexes[indexLine].Line]
+		partBuf := buf[indexes[indexLine].Pos:]
+		part2 := string(([]rune(partBuf))[:indexPos])
+		posInGrapheme = stringWidth(part2)
+	}
+
+	// store last cursor position
+	lastCy := f.cursor.y
+
+	// cursor must be inside screen
+	x, y, width, height := f.GetInnerRect()
+	_ = width
+	f.cursor.x = posInGrapheme + x - f.view.columnOffset
+	f.cursor.y = indexLine + y - f.view.lineOffset
+	if isOutsideBuffer {
+		f.cursor.x++
+	}
+	if y+height <= f.cursor.y {
+		diff := (indexLine - lastIndexLine) - (y + height - lastCy) + 1
+		f.view.lineOffset += diff
+	}
+	if f.cursor.y < y {
+		diff := -((lastIndexLine - indexLine) - lastCy) - 1
+		f.view.lineOffset += diff
+	}
+	// TODO: other cases with columnOffset
+}
+
+// InputHandler returns the handler for this primitive.
+func (f *TextArea) InputHandler() func(event *tcell.EventKey, setFocus func(p Primitive)) {
+	return f.WrapInputHandler(func(event *tcell.EventKey, setFocus func(p Primitive)) {
+		// Moving strategy
+		//
+		//	* Moving only in buffer coordinate, not on screen coordinate
+		//	* Move up/down - moving by buffer lines
+		//	* Move left/right - moving by buffer runes
+		//
+		line, pos := f.cursorByScreen()
+		key := event.Key()
+		switch key {
+		case tcell.KeyUp:
+			if line <= 0 {
+				// do nothing
+			} else {
+				line--
+			}
+		case tcell.KeyDown:
+			if len(f.view.buffer)-1 <= line {
+				// do nothing
+			} else {
+				line++
+			}
+		case tcell.KeyLeft:
+			if pos == 0 {
+				// do nothing
+			} else {
+				pos--
+			}
+		case tcell.KeyRight:
+			if len(f.view.buffer) == 0 {
+				break
+			}
+			if stringWidth(f.view.buffer[line]) <= pos {
+				// do nothing
+			} else {
+				pos++
+			}
+		case tcell.KeyHome:
+			pos = 0
+		case tcell.KeyEnd:
+			if 0 <= line && line < len(f.view.buffer) {
+				pos = len([]rune(f.view.buffer[line]))
+			}
+		case tcell.KeyEnter:
+			f.insertNewLine()
+			line++
+			pos = 0
+		case tcell.KeyDelete:
+			pos++
+			defer func() {
+				f.deleteRune()
+				pos--
+				f.cursorByBuffer(line, pos)
+			}()
+		case tcell.KeyBackspace, tcell.KeyBackspace2:
+			line, pos = f.deleteRune()
+		default:
+			r := event.Rune()
+			if unicode.IsLetter(r) ||
+				unicode.IsNumber(r) ||
+				unicode.IsMark(r) ||
+				unicode.IsSpace(r) {
```

Comment:
This filtering is too strict - it blocks user from entering any special characters, eg !"£$%^&*(){}~@:<>?  and more. I think it could be reduced to just  "unicode.IsGraphic", so that it only blocks control sequences.

Reply:
I agree with you.